File: src/main/java/com/zndevs/toolset/commands/ToolSetUpdateCommand.java
package com.zndevs.toolset.commands;

import com.zndevs.toolset.ToolSetCommand;
import com.zndevs.toolset.ToolSetConfiguration;
import com.zndevs.toolset.tools.CommandResponse;
import com.zndevs.toolset.tools.ToolSetOptions;
import com.zndevs.toolset.tools.URLDownloader;
import com.zndevs.toolset.updater.CommandUpdateVersions;
import com.zndevs.toolset.updater.ToolSetUpdater;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.plugin.InvalidDescriptionException;
import org.bukkit.plugin.InvalidPluginException;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.java.JavaPlugin;
import org.json.simple.parser.ParseException;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

/**
 * ToolSet Update command - Updates all official ToolSet modules
 */

public class ToolSetUpdateCommand implements ToolSetCommand {

    @Override
    public boolean onCommand(CommandSender sender, Command command, String s, String[] args) {
        CommandResponse response = new CommandResponse();

        if (args.length == 0) {
            response.setPlayerMessage(ToolSetOptions.getHelpFormat("ToolSet Updater", new String[]{
                    "check, chk, c",
                    "update, upd, u"
            }));
        } else {
            String arg = args[0];
            switch (arg) {
                case "check":
                case "chk":
                case "c":
                    response = checkForUpdates(sender, response);
                    break;
                case "update":
                case "upd":
                case "u":
                    response = doUpdate(sender, response);
                    break;
                default:
                    response.setPlayerMessage(ToolSetOptions.getHelpFormat("ToolSet Updater", new String[]{
                            "check, chk, c",
                            "update, upd, u"
                    }));
                    break;
            }
        }

        if (response.usePlayerMessage()) {
            sender.sendMessage(response.getPlayerMessage());
        }
        return false;
    }

    @Override
    public ToolSetConfiguration setConfigurationDefaults(ToolSetConfiguration defaults) {
        return defaults;
    }

    private CommandResponse checkForUpdates(CommandSender sender, CommandResponse response) {
        if (sender.hasPermission("toolset.core.admin.update.check")) {
            ToolSetUpdater updater = new ToolSetUpdater();
            try {
                List<CommandUpdateVersions> updatesRequired = updater.checkForUpdates();
                StringBuilder updates = new StringBuilder(ToolSetOptions.prefix + "Available updates:");
                for (CommandUpdateVersions currentVersion : updatesRequired) {
                    updates.append("\n")
                            .append(currentVersion.getCommandName())
                            .append("(v")
                            .append(currentVersion.getOldVersion())
                            .append(") can be updated to v")
                            .append(currentVersion.getNewVersion());
                }
                response.setPlayerMessage(updates.toString());
            } catch (IOException e) {
                response.setPlayerMessage(ToolSetOptions.errorPrefix + "An unknown error occurred while checking for updates. Please check the console for more information.");
                e.printStackTrace();
            } catch (ParseException e) {
                response.setPlayerMessage(ToolSetOptions.errorPrefix + "An unknown error occurred while parsing the updater data. Please check the console for more information.");
                e.printStackTrace();
            }
        } else {
            response.setPlayerMessage(ToolSetOptions.noPermissionsMessage);
        }

        return response;
    }

    private CommandResponse doUpdate(CommandSender sender, CommandResponse response) {
        ToolSetUpdater updater = new ToolSetUpdater();

        try {
            boolean areUpdatesRequired = updater.updatesRequired();

            if (!areUpdatesRequired) {
                response.setPlayerMessage(ToolSetOptions.prefix + "No updates required");
                return response;
            }
        } catch (IOException | ParseException e) {
            response.setPlayerMessage(ToolSetOptions.errorPrefix + "An unknown error occurred while checking for updates. Please check the console for more information.");
            e.printStackTrace();
        }

        if (sender.hasPermission("toolset.core.admin.update") || sender.hasPermission("toolset.core.admin.update.update")) {
            URLDownloader downloader = new URLDownloader();
            try {
                // Get updates required
                List<CommandUpdateVersions> updatesRequired = updater.checkForUpdates();
                StringBuilder updates = new StringBuilder(ToolSetOptions.prefix + "Available updates:");
                for (CommandUpdateVersions currentVersion : updatesRequired) {

                }
                response.setPlayerMessage(updates.toString());

                String updaterUrl = ToolSetOptions.getToolSetApiPath("files/ToolSetUpdater-1.0.0.jar");
                System.out.println("Downloading " + updaterUrl);
                byte[] updaterBytes = downloader.downloadUrlBytes(updaterUrl);

                // Create temporary file
                Path tempFilePath = new File(ToolSetOptions.getDataDirectory().getParentFile(), "toolset.updater.jar").toPath();
                Files.write(tempFilePath, updaterBytes);

                // Also create the information file
                File infoFile = new File(
                        ToolSetOptions.plugin.getDataFolder().getParent(),
                        "updates.toolset.dat"
                );
                try (PrintWriter output = new PrintWriter(infoFile)) {
                    for (CommandUpdateVersions update : updatesRequired) {
                        JavaPlugin plugin = updater.getPlugin(update);
                        if (plugin != null) {
                            String pluginName = updater.getPluginInternalName(update);
                            output.print(pluginName + "\n");
                        }
                    }
                }

                Plugin updaterPlugin = Bukkit.getPluginManager().loadPlugin(tempFilePath.toFile());
                Bukkit.getPluginManager().enablePlugin(updaterPlugin);
            } catch (IOException e) {
                response.setPlayerMessage(ToolSetOptions.errorPrefix + "An unknown error occurred while downloading the updater. Please check the console for more information.");
                e.printStackTrace();
            } catch (InvalidDescriptionException | InvalidPluginException e) {
                response.setPlayerMessage(ToolSetOptions.errorPrefix + "An unknown error occurred while enabling the updater. Please check the console for more information.");
                e.printStackTrace();
            } catch (ParseException e) {
                response.setPlayerMessage(ToolSetOptions.errorPrefix + "An unknown error occurred while getting the required updates. Please check the console for more information.");
                e.printStackTrace();
            }
        } else {
            response.setPlayerMessage(ToolSetOptions.noPermissionsMessage);
        }

        return response;
    }
}

File: src/main/kotlin/com/zndevs/toolset/commands/ToolSetUpdateCommand.kt
package com.zndevs.toolset.commands

import com.zndevs.toolset.ToolSetCommand
import com.zndevs.toolset.ToolSetConfiguration
import com.zndevs.toolset.commands.tools.SubCommand
import com.zndevs.toolset.tools.CommandResponse
import com.zndevs.toolset.tools.ToolSetOptions
import com.zndevs.toolset.tools.URLDownloader
import com.zndevs.toolset.updater.ToolSetUpdater
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.plugin.InvalidDescriptionException
import org.bukkit.plugin.InvalidPluginException
import org.json.simple.parser.ParseException
import java.io.File
import java.io.IOException
import java.io.PrintWriter
import java.nio.file.Files

class ToolSetUpdateCommand : ToolSetCommand() {
    private val subCommands: Array<SubCommand> = arrayOf(
            SubCommand("check", arrayOf("chk", "c"), "Check for updates", fun(s, _, _, _, _, r): Boolean {
                checkForUpdates(s, r)
                return true
            }),
            SubCommand("update", arrayOf("upd", "u"), "Update if required", fun(s, _, _, _, _, r): Boolean {
                doUpdate(s, r)
                return true
            })
    )

    private fun checkForUpdates(sender: CommandSender, response: CommandResponse) {
        if (!sender.hasPermission("toolset.core.admin.update.check")) {
            response.playerMessage = ToolSetOptions.noPermissionsMessage
            return
        }

        val updater = ToolSetUpdater()
        try {
            val updatesRequired = updater.checkForUpdates()
            val updates = StringBuilder(ToolSetOptions.prefix).append("Available updates:")

            for (currentVersion in updatesRequired) {
                updates.append("\n")
                        .append(currentVersion.commandName)
                        .append(" (").append(currentVersion.oldVersion).append(")")
                        .append(" can be updated to ").append(currentVersion.newVersion)
            }

            if (updatesRequired.isEmpty()) updates.append(" &aToolSet is all up to date.")
            response.playerMessage = updates.toString()
        } catch (e: IOException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while checking for updates. Please refer to the console for more information."
            e.printStackTrace()
        } catch (e: ParseException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while parsing the updater data. Please refer to the console for more information."
        }
    }

    private fun doUpdate(sender: CommandSender, response: CommandResponse) {
        if (!sender.hasPermission("toolset.core.admin.update") && !sender.hasPermission("toolset.core.admin.update.update")) {
            response.playerMessage = ToolSetOptions.noPermissionsMessage
            return
        }

        val downloader = URLDownloader()
        val updater = ToolSetUpdater()

        try {
            // check for updates
            val updatesRequired = updater.updatesRequired()

            if (!updatesRequired) {
                response.playerMessage = ToolSetOptions.prefix + "No updates are required"
                return
            }
        } catch (e: IOException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while checking for updates. Please refer to the console for more information."
            e.printStackTrace()
        } catch (e: ParseException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while parsing the updater data. Please refer to the console for more information."
        }

        try {
            // get a list of required updates
            val requiredUpdates = updater.checkForUpdates()

            val updaterUrl = ToolSetOptions.getToolSetApiPath("files/ToolSetUpdater-1.0.0.jar")
            val updaterBytes = downloader.downloadUrlBytes(updaterUrl)

            // create temporary updater file
            val tempUpdaterPath = File(ToolSetOptions.dataDirectory.parentFile, "toolset.updater.jar")
            Files.write(tempUpdaterPath.toPath(), updaterBytes)

            // also create information file to say what to update
            val infoFile = File(ToolSetOptions.plugin?.dataFolder, "updates.toolset.dat")

            PrintWriter(infoFile).use {
                for (update in requiredUpdates) {
                    val plugin = updater.getPlugin(update)
                    if (plugin != null) {
                        val pluginName = updater.getPluginInternalName(update)
                        it.print(pluginName + "\n")
                    }
                }
            }

            val updaterPlugin = Bukkit.getPluginManager().loadPlugin(tempUpdaterPath)
            Bukkit.getPluginManager().enablePlugin(updaterPlugin)
        } catch (e: IOException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while downloading the updater. Please refer to the console for more information."
            e.printStackTrace()
        } catch (e: InvalidDescriptionException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while enabling the updater. Please refer to the console for more information."
            e.printStackTrace()
        } catch (e: InvalidPluginException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while enabling the updater. Please refer to the console for more information."
            e.printStackTrace()
        } catch (e: ParseException) {
            response.playerMessage = ToolSetOptions.errorPrefix + "An error occurred while getting the required updates. Please refer to the console for more information."
        }
    }

    override fun getSubCommands() = subCommands

    override fun getCommandName() = "ToolSet Updater"

    override fun setConfigurationDefaults(defaults: ToolSetConfiguration) = defaults
}